1.fee_bumping/fee_bumping.md

```diff
@@ -177,21 +179,128 @@ TODO
+of an unconfirmed (_parent_) transaction as an input to a new (_child_)
+transaction. The wallet attaches enough fee to the child transaction to
+increase the combined feerate across the parent and child transactions.
+
+## How does CPFP work?
+
+When constructing a new block, miners are incentivized to fill the 1vMB with
+the set of transactions that maximize their miner fee. If all unconfirmed
+transactions were independant, this would be a very straightforward operation -
+the miner would select the transaction with the highest feerate and add it to
+the candidate block. She'd then take the transaction with the next highest
+feerate and add it to the block. She'd continue to do this until the block was
+full. This trivially maximizes her profit from the block (with a little
+complication around the final few bytes of the block to ensure that she'd
+maximally filled the block).
+
+However, unconfirmed transactions _aren't_ independant. It is possible to have
+chains of unconfirmed transactions by spending the output from a transaction
+before it is included in a block. For example, if tx A has two outputs a1 and
+a2, transaction B could use one of those outputs as an input before A has
+been included in a block. In this case, if the miner wants to include
+transaction B in the block, she must also include transaction A, since without
+A, B is spending a non-existent output and is invalid.
+
+If a miner considered transactions independently when constructing her block,
+she may forego transactions with very high fees if they depended on
+transactions with very low fees (or worse, she may construct an invalid block
+with a transaction that depends on an unincluded transaction). To maximize her
+profit, the miner should therefore consider transactions in _packages_ (sets of
+transactions with dependencies on each other) when constructing a new block.
+
+Wallets can take advantage of this rational behavior by miners to incentivize
+them to include a stuck, low-fee transaction, by spending one of its outputs
+and increasing the total feerate across the transaction package.
+
+## History of CPFP
+
+For users to be able to bump a transaction using CPFP, two elements are required:
+
+- a wallet that will spend the output of a stuck unconfirmed transaction in
+  order to bump the combined feerate.
+- The expectation that miners will maximize their profit by considering
+  packages of transactions.
+
+Until 2012, the Bitcoin Core mining component did not consider fee-rate when
+ordering transactions for inclusion in a block. Until then, transactions had
+been ordered by 'prority' (the sum of the (value X coin age) for each
+transaction input, divided by the transaction size). Bitcoin Core [PR #1590][]
+changed the mining code to sort transactions by fee rate. [Version 0.7.0][],
+released in September 2012 was therefore the first Bitcoin Core release with
+any fee-optimized behavior for miners.
```

Comment:
AFAIK, the preceeding text about ordering transactions by fee is correct, however I think you're using that information to draw the conclusion that 0.7 was the first release with fee-optimized behavior---which I think is an incorrect conclusion.

Nakamoto introduced code in 0.3.something, I believe, that filled the block template in priority order (as described) but it also increased the minimum fee for a transaction to be added to the block template as the template approached half of what was called then the "soft maximum block size" (which was 500KB, so 250KB was half).  The increases were exponential, so e.g. the first 100KB could be cheap or free but the last 50KB could be quite pricey.

This is not fee optimal---for that you need feerate sorting with CPFP as we have now[1]---but it is fee optimized.  If you need, I can find that Nakamoto code.

[1] Although to be truly optimal, you probably need to have infinitely spacious mempools and consider CPFP tx chains of infinite depth.  Happily, good approximations of optimal can be achieved much more simply.

Reply:
You're good at this! Yes, Satoshi added code here to raise the fee required as the block filled up: https://github.com/bitcoin/bitcoin/commit/f1e1fb4bdef878c8fc1564fa418d44e7541a7e83#diff-23cfe05393c8433e384d2c385f06ab93R584, and then updated it here: https://github.com/bitcoin/bitcoin/commit/9b8eb4d6907502e9b1e74b62a850a11655d50ab5#diff-23cfe05393c8433e384d2c385f06ab93R587.

You're right about the current strategy being optimized, but not optimal. Maximizing the fee yield in a block is a tricky problem. Even if there we knew a better-than-brute-force method to make transaction selection optimal, it'd need to be efficient. There's not much use in an algorithm that finds the optimal block if it takes ten minutes to do so!